feat(LoginPage): add LoginPage component with storybook and tests.

Laviro opened the pull request:
affects: patternfly-react

Add LoginPage component with storybook and tests. fix #80

[**Pattern**](http://www.patternfly.org/pattern-library/application-framework/login-page/)
[**Storybook**](https://rawgit.com/Laviro/patternfly-react/storybook/feature/login-page/index.html?knob-Social%20List%20Size=12&selectedKind=patternfly-react%2FApplication%20Framework%2FLogin%20Page&selectedStory=Managed%20Basic%20Login%20Page&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

**Basic Login Page**:
![6a3ec2f0-6c01-46eb-y62e-76804af3b265](https://user-images.githubusercontent.com/26363699/40704377-925a526e-63f0-11e8-8b89-9d4ba7dd0e22.gif)


**Social Login Page**
![8b73673d-1a5f-425d-y2d5-eb0a414b715e](https://user-images.githubusercontent.com/26363699/40704641-5b584ee6-63f1-11e8-8d71-700ef43e1887.gif)

jeff-phillips-18 commented:
Theoretically, if we mimic the HTML from http://www.patternfly.org/pattern-library/application-framework/login-page/#code#login-layout-markup we shouldn't need additional less/sass.

Laviro commented:
@jeff-phillips-18 I added some styles to make the page less "jumpy" when the page change dynamically.

lizsurette commented:
@Laviro This is looking super great. I'm excited to use this in my product :)

I just did a quick review from a UX point of view (FYI @serenamarie125 @jgiardino) and noticed a few small things that I think would be great to update as you continue working on this.

1) When there aren't any errors, the padding of the form elements seems to be a bit too much. (Between email address and password fields, and remember me checkbox. And then again between the log in button and need an account link)

2) In the PF base implementation example, the language that is selected (English in this case) is actually highlighted as selected when the user drops down the menu. Is this something we can do here?

3) A few little typo/text nitpicks: Updates should be made that we use a lower case “a” in "Email address" and uppercase "I" in "Log In".

4) I like the way you've broken this up into 3 sections in storybook. Any chance the section names could be updated to be a bit clearer for each use case? Maybe “Basic Login Page”, “Login Page with Errors”, and “Login Page with Social Options”? Totally open to other suggestions, just thought this would be clearer.

5) On the social options page, the less link needs an updated icon pointing up rather than down.

6) On the social options page, the responsive state is missing the separator bar between the log in button and social options.

Great work on this! I hope this feedback helps.

Liz

Laviro commented:
@lizsurette Thanks for the feedback :)
I will try to answer to some of your thoughts.

1. Input errors: I agree that the padding are big, but as I have added validation to the form and errors have started to appear, the interaction seems too annoying..
You can try to press the CapsLock while in the password field or submit the form while empty, and than type something else in the inputs, and submit again.. I can make another version where the errors will collapse and you can see the difference.

2. Language Picker dropdown highlight - no problem, we can add it !

3. Letter-Case updates - will be fixed.

4. Rename storybooks: Thanks, I guess this names need to be clearer :)

5. Arrow Up on "Less" button - wil be fixed

6. Seperator bar - will be fixed

I guess you will see my updates soon,  I would be happy to know what you think about the input's errors.

Thanks, Ron

lizsurette commented:
@mcarrano @Laviro Yeah, I see the error cases in the Design tab: http://www.patternfly.org/pattern-library/application-framework/login-page/#design but I don't see them anywhere in the Code sample.

I've logged a design issue [here](https://github.com/patternfly/patternfly-design/issues/671) so that we can continue this discussion.

Thanks,
Liz

jeff-phillips-18 commented on packages/core/less/login-page.less:
```diff
@@ -0,0 +1,42 @@
+
+.login-pf-brand {
+  margin-top: 0 !important;
+}
```
I don't think we should change this. PF supplies the desired margin top. The issue is only when there is an inline notification above that this makes sense. Could we modify the inline notification to be positioned within this margin? Currently when you dismiss the notification the page jumps up.

Laviro replied:
I can change the alert to be absolute and than the page won't jump, 
although it will hide the page header on smaller screens or when there's a lot of text in the alert.
Not sure what is the best option, @lizsurette we need your help here.

**Alert's position is absolute**
![a3da674a-8920-45a5-y214-b252df854615](https://user-images.githubusercontent.com/26363699/40785861-4d41d460-64f2-11e8-96e0-cf28d8c6ef5d.gif)

**Alert's display is block**
![15cb04bf-1067-42a0-yadc-d4b48a205095](https://user-images.githubusercontent.com/26363699/40786410-a6b73d2c-64f3-11e8-95a9-cb0c05f49683.gif)

Laviro commented on packages/core/less/login-page.less:
```diff
@@ -0,0 +1,42 @@
+
+.login-pf-social-section {
+  min-height: 335px;
+}
```
I think that the card jump too much without it:
![75a67192-14f0-4cb5-y231-e487a9c4385d](https://user-images.githubusercontent.com/26363699/40781060-3c50a776-64e3-11e8-836a-0279b08ae00e.gif)

**With min-height:**
![cfc05517-9912-428e-y759-00955da28e06](https://user-images.githubusercontent.com/26363699/40781099-6129bb28-64e3-11e8-8cf1-0ff85c0da385.gif)

jgiardino replied:
I added a [question about this behavior](https://github.com/patternfly/patternfly-design/issues/671#issuecomment-393531758) to the issue that @lizsurette opened about the error messages. 

(I posted this in the wrong comment box earlier 😁)

jeff-phillips-18 replied:
I guess it depends on the number of social links there are.

Laviro replied:
Maybe it will feel better if we'll move the sign-up section to be under the submit button,
what do you think?

Laviro commented on packages/core/less/login-page.less:
```diff
@@ -0,0 +1,42 @@
+  background-size: cover;
+  background-position: center;
+  padding: 20px;
+}
```
The background breaks when `background-size: auto;`
Without `padding` there's not enough space between the page edges.
Although `background-position` && `background-repeat` 
can be removed.

jgiardino replied:
I added a [question about this behavior](https://github.com/patternfly/patternfly-design/issues/671#issuecomment-393531758) to the issue that @lizsurette opened about the error messages.

jeff-phillips-18 commented:
@lizsurette Have we any decisions on the ux questions above? Can we do a first pass here and iterate as decisions are finalized?

jeff-phillips-18 commented:
@patternfly/patternfly-react-ux This PR is awaiting some UX feedback. I'd like to move forward and we can iterate as necessary but need some preliminary decisions.

Issue 1: Resize on inline notifications?
Issue2: Set a minimum height on the social section so as not to resize, move the sign-up section to be under the submit button?

jgiardino commented:
My suggestion would be to consume the patternfly css from core as it's currently defined, until we have updates on design from the patternfly design team. I think in doing this, it would also be easier for the designers to review the issues that were noted and provide better guidance on how to resolve them.

lizsurette commented:
@jeff-phillips-18 @Laviro Sorry for the delay on my end! After a discussion with the PatternFly core team I agree with @jgiardino that we should move forward with consuming the PatternFly core css for both of these issues. We did have a good conversation about the benefits of allowing for space in the login form to show errors without having the form jump a bit, and agreed it will be considered for the next version of PatternFly, but there was some pushback on doing it for PatternFly 3 which was noted in the issue [here](https://github.com/patternfly/patternfly-design/issues/671).

I hope this helps clarify!

gregsheremeta commented:
Hi!

This is great. I have a question.

I haven't studied the code yet and I'm a react novice. In oVirt, we have one extra form field called 'profile' which roughly means 'which internal sso to authenticate to'

Can the component accommodate that?

Laviro commented:
Hey @gregsheremeta ! 
At first I also thought that allowing to add extra fields is a great idea..
But then.. do we want to add it before/after/to the side of the current inputs ?
There are so many use-cases, and I guess we can't support it all,

Though you can copy the component structure from the storybook 
and replace or change the `LoginCardForm` component with whatever form you want :)

@patternfly/patternfly-react-ux Do you have any thoughts about this?

gregsheremeta commented:
thank you @Laviro 
for what it's worth, in oVirt we put it below (haven't updated to the latest login design yet obviously :))

(I never loved the word 'profile' there ...)

![selection_016](https://user-images.githubusercontent.com/2248293/41623905-2b25c13a-73e2-11e8-892d-5dab19f044a5.png)

Laviro commented:
@gregsheremeta 
Thanks for uploading the oVirt pic here,
lets wait to see what the ux-team think.

lizsurette commented:
@gregsheremeta @Laviro I did some digging in the [Design write-up for the pattern](http://www.patternfly.org/pattern-library/application-framework/login-page/#design) and additional fields are called out that they should live directly below the password field:

"Required Fields: Username and password are the required fields while additional fields are optional. Some applications may require additional fields such as a specific server or geo-location. When additional fields are required, they should be ordered on a case-by-case basis and placed directly below the password field. The login card will expand to accommodate the additional fields. Try to limit the total number of fields to less than five. Also, you may use progressive disclosure if the field value have dependency on the previous relevant field."

Laviro commented:
Thanks @lizsurette, I totally missed this part..
going to add this ability to insert optional inputs,
and thanks @gregsheremeta for bringing up this issue !!

jeff-phillips-18 commented:
I think it would make sense to continue as is and add the optional additional fields as a follow on.

lizsurette commented:
> I think it would make sense to continue as is and add the optional additional fields as a follow on.

I'm fine with that!

jgiardino commented:
@Laviro Let us know if there is anything blocking you from moving forward with this one.

Laviro commented:
@jgiardino I think we can continue with code reviews, 
and add that feature in one of the next commits or as a fix.

jgiardino commented:
Sounds good. I was waiting to do the html/css review after the custom css was removed.

jeff-phillips-18 commented:
@Laviro Do you plan to make the changes to remove the custom CSS?

Laviro commented:
@jgiardino I have fixed the HTML structure in the `BasicLoginPage`to be the same as in the pattern,
removed the custom-css and updated the [storybook](https://rawgit.com/Laviro/patternfly-react/storybook/feature/login-page/index.html?selectedKind=patternfly-react%2FWidgets%2FLogin%20Page&selectedStory=Build%20Your%20own%20Basic%20Login%20Page&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs).

jeff-phillips-18 commented on packages/patternfly-react/src/components/LoginPage/LoginPage.stories.js:
```diff
@@ -0,0 +1,223 @@
+import { name } from '../../../package.json';
+
+const stories = storiesOf(
+  `${storybookPackageName(name)}/${STORYBOOK_CATEGORY.WIDGETS}/Login Page`,
```
LoginPage should be under `${STORYBOOK_CATEGORY.APPLICATION_FRAMEWORK}`

Laviro commented on packages/patternfly-react/src/components/LoginPage/components/LoginCardComponents/LoginCardWithValidation.js:
```diff
@@ -0,0 +1,254 @@
+
+  onInputKeyUp = (e, inputType) => {
+    this.props[inputType].onKeyUp && this.props[inputType].onKeyUp(e);
+    this.onCapsLockAsModifier(e);
```
I removed this function as I wasn't even use it after all..
The CapsLock warning updates by the mouseEnter & the window - keyup events.

It works for me well, I am using Fedora and chrome v66.
![aug 8 2018 4_34 pm](https://user-images.githubusercontent.com/26363699/43840315-11d769cc-9b29-11e8-9e83-551293ae7737.gif)

I saw that there are problems with CapsLock detection on Mac and some browser,
going to check for a solution that will match most users.

Laviro commented:
@jgiardino Do you have any comments for the Login page structure or style?
The problem that I've faced are:
- Login page background style
- Login card input height ( for solving the jumping errors/warnings over an input )
- Login form error ( color and margin )
- Social Login divider ( margin and border )
- Social Login columns ( margin on small screens )
- page alert ( margin )

Laviro commented:
@gregsheremeta @lizsurette I have added the `additionalFields` prop 
for inserting more fields under the password field.

jgiardino commented:
Hi @Laviro, thanks again for contributing this! You've done a great job implementing one of our more complex components! And you've added a lot of nice touches. 👌 

> Login page background style

When I look at the basic login page using dev tools, I see that there's a background image defined as an inline style:
`<div class="login-pf" style="background-image: url(&quot;static/media/login-screen-bg.bb6b114e.jpg&quot;);">`

But if I disable that style, it looks like there's also a background image defined for `.login-pf` that displays. Is there a reason why the background image is added as an inline style? Can we remove that, or were there other issues related to this?

> Login card input height ( for solving the jumping errors/warnings over an input )

My understanding from past discussions about this is that the implementation as it is now works as expected. I would love to see forms handle error messages in a more graceful way, but I think implementing that type of feature in a one-size-fits-all way that accommodates translation and the possibility of multiple error cases per field is not an easy thing to solve. 

> Login form error ( color and margin )

What you have looks correct. 

> Social Login divider ( margin and border )
> Social Login columns ( margin on small screens )

When you click the "More" link to show more options, there's a larger margin between the options that were visible and the options that become visible. 

![image](https://user-images.githubusercontent.com/21063328/43855571-08766b54-9b14-11e8-9d2d-7d460b4674b6.png)

Instead of having two separate `ul.login-pf-social.list-unstyled.login-pf-social-double-col` elements, can there just be one, with the extra `<li>` elements added to the original one? In the core html/css example there's an additional class `login-pf-social-all` on the `<ul>` that will hide the extra options. But I don't think it's a big deal if that solution doesn't work with this implementation, because I think we need to modify the margin anyway...

The core html/css implementation doesn't include the divider, but the design does. I think for now, we can assume that the design is correct, and keep the divider like you have it, but there aren't styles in the core css to produce the same layout that's shown in the design. And right now, the spacing in this implementation doesn't fully match the design:
![image](https://user-images.githubusercontent.com/21063328/43861767-ced1c79c-9b25-11e8-88d3-5ccfe82a5f3d.png)

My suggestion would be to add the following to the pf-react css, which will also fix the margin issue that I mention above: 
    `.login-pf-page .login-pf-social {margin-top: 0;}`

This will not match the design document exactly, but I think it's close enough, and we need to create a separate issue anyway to track down why there's a difference and figure out what the expected solution should be. So we can revisit this layout later if needed when that issue is resolved.


Also in this implementation, I see that you animate the transition when clicking the Less/More link, which is a really nice touch. But it seems to stick when expanding. It shows the first 3, pauses, then expands all the way to show the 4th one. Do you see that in your build? Is there anyway to adjust that animation?

> page alert ( margin )

I can't find an example of this in core. I'd say it's fine as it is for now, and we can revisit this element in core and update later as needed.

One other thing I noticed - 
The warning that displays for the caps lock key doesn't always display when expected. When I have focus in the password field, and toggle the caps lock key, keypress event seems to be captured later than it should.

Laviro commented:
Wow @jgiardino, thanks for the detailed review !

> Is there a reason why the background image is added as an inline style?

It gives a way for the developers to insert a background from props without dealing with css:
![selection_039](https://user-images.githubusercontent.com/26363699/43900186-c553f100-9bec-11e8-9213-2af4a0553ef8.png)

 > it looks like there's also a background image defined for .login-pf that displays. 

I think pf should remove this image from the `.login-pf` since it won't fit every product. 

> instead of having two separate `ul.login-pf-social.list-unstyled.login-pf-social-double-col` elements, can there just be one, with the extra `<li>` elements added to the original one? 

I think there are three options:
- if we want to keep the animation: 
- leave it just as it is right now with two classes.
- remove the collapsing div, and try to create collapse animation by css rules - not easy.
- remove the collapse.

providing gifs of the options:
**Regular**
![aug 9 2018 4_26 pm](https://user-images.githubusercontent.com/26363699/43902619-3e99dbb4-9bf3-11e8-99ed-e81ffce93f16.gif)

**Collapsed**
![aug 9 2018 4_27 pm](https://user-images.githubusercontent.com/26363699/43902643-4830eaa0-9bf3-11e8-9d83-c3bdba0c32c8.gif)

> The warning that displays for the caps lock key doesn't always display when expected. 

I was writing about this issue here above, it works fine for me,
I am using Fedora and chrome v66.
![aug 8 2018 4_34 pm](https://user-images.githubusercontent.com/26363699/43840315-11d769cc-9b29-11e8-9e83-551293ae7737.gif)

I saw that there are problems with CapsLock detection on Mac and some browsers,
going to check for a solution that will match most users.

jgiardino commented:
Thanks for the clarification about the background image. That makes a lot of sense! 😉 I opened an issue in patternfly core to capture this question, plus a couple of others that came up during review: https://github.com/patternfly/patternfly/issues/1125

I was just chatting with @mcarrano about the issues with the responsive state (i.e. the issue with the extra margin and the issue with the animation not being a smooth transition from start to finish when clicking More). @mcarrano's recommendation for resolving this is to remove the More/Less toggle from the responsive state, since there isn't much benefit to the user in collapsing the extra options on a mobile device. 

Is it possible to remove the toggle from the responsive state and keep it for the other? @mcarrano's feedback was that if it's problematic to do this, then his recommendation is to remove the toggle altogether.

> I saw that there are problems with CapsLock detection on Mac and some browsers,
going to check for a solution that will match most users.

If you are having difficulty reproducing this issue and finding a solution for it, we can capture this in a follow-up issue for someone with a mac to look into.

matthewcarleton commented:
hey @Laviro! This is looking great. Thanks for putting in all the effort here. I'm going to do the final CSS review so we can get this moving while Jenn is away on PTO next week. 
In addition to what she pointed out above I noticed one other thing. 
The language selector is not aligned to the right like we have in the core example. From looking at the code in the react implementation it looks like a slight difference in how it was built.  It just a matter of adding `<span class="filter-option pull-left">` around the "english text and adding `.btn-group` to the wrapping div with `.bootstrap-select`. If this isn't clear just let me know!

Laviro commented:
@matthewcarleton @jgiardino 
thank you for your comments ! I fix the styling problems mentioned above,
can we continue with this PR and adjust the styles after the issues you've created will be closed?

@jeff-phillips-18 I think I've fixed the `capslock` behavior, can you check if it works on mac ? thanks :)

matthewcarleton commented:
Hey @Laviro, great work, thanks!
I'd suggest pulling out the <hr> and that should resolve the margin issue. The line for mobile should not be in the design and the design will be updated to reflect this :)

What are your thoughts on removing the more/less toggle? Is this difficult to do? I like @mcarrano's suggestion of removing it altogether too if that more reasonable from a technical stand point.

I think if we address those two things we can move forward with this.

priley86 commented:
This is looking really nice to me @Laviro ! I will add a couple of comments inline, but overall very happy w/ this.

One thing you will definitely want to do before merging is export your Login components in the main patternfly-react barrel file (`patternfly-react/src/index.js`). It would be nice to see all of your individual components/patterns exported via a component index file (`components/LoginPage/index.js`). See Table [index.js](https://github.com/patternfly/patternfly-react/blob/master/packages/patternfly-react/src/components/Table/index.js) for example.

priley86 commented on packages/patternfly-react/src/components/LoginPage/assets/img/logos/index.js:
```diff
@@ -0,0 +1,27 @@
+import dropbox from './dropbox-logo.svg';
```
we do not need to duplicate these logos here in patternfly-react. They can be consumed/re-exported directly from Patternfly since we are directly dependent on PF.

That being said, I think we can change this to : 
```
import dropbox from 'patternfly/dist/img/dropbox-logo.svg';
import facebook from 'patternfly/dist/img/facebook-logo.svg';
import fedora from 'patternfly/dist/img/fedora-logo.png';
import git from 'patternfly/dist/img/git-logo.svg';
import github from 'patternfly/dist/img/github-logo.svg';
import google from 'patternfly/dist/img/google-logo.svg';
import instagram from 'patternfly/dist/img/instagram-logo.png';
import linkedin from 'patternfly/dist/img/linkedin-logo.svg';
import openID from 'patternfly/dist/img/open-id-logo.svg';
import skype from 'patternfly/dist/img/skype-logo.svg';
import stackExchange from 'patternfly/dist/img/stack-exchange-logo.svg';
import twitter from 'patternfly/dist/img/twitter-logo.svg';
```

and delete all of the new logo assets coming from PF...

priley86 commented on packages/patternfly-react/src/components/LoginPage/components/LoginCardComponents/LoginCardWithValidation.js:
```diff
@@ -0,0 +1,253 @@
+    if (!this.state.passwordField.value) {
+      return;
+    }
+    e.key === 'CapsLock' &&
```
We started tracking all keycodes for accessibility in `src/helpers.js` , `KEY_CODES` constant. I think it is best to have one constant for these.

priley86 commented on packages/patternfly-react/src/components/LoginPage/components/LoginCardComponents/LoginCardWithValidation.js:
```diff
@@ -0,0 +1,253 @@
+    const keyCode = e.keyCode ? e.keyCode : e.which;
+    const shiftKey = e.shiftKey ? e.shiftKey : keyCode === 16;
+    const isCapsLock =
+      (keyCode >= 65 && keyCode <= 90 && !shiftKey) ||
```
(same as above)

priley86 commented on packages/patternfly-react/src/components/LoginPage/LoginPage.js:
```diff
@@ -0,0 +1,87 @@
+import React from 'react';
```
Definitely a higher order component (pattern). The `LoginPageWithTranslation` and `LoginCardWithValidation` components seem to do a lot of internal state management. I think I would generally handle that in the application, but since you have done such a nice job breaking down the components, this should not be an issue (and will probably help a lot of consumers needing less fine grain detail). I'm OK proceeding here if others are ;-)

Laviro replied:
Thank you so much @priley ! 
The consumer of this component can decided whether they 
want the component with validation or not,
by setting the `form: { validate }` prop.

matthewcarleton commented:
awesome! @jeff-phillips-18 this is good to go from a CSS perspective.

dgutride commented:
@laviro, @LHinson, from talking with  @priley86, we have two necessary changes:
I've 👍  them above (removing logos and fixing the barrel file to include the necessary exports).  After that, we can track the keycode changes using a separate issue.

Please add a comment and tag @priley86 and I once those two changes have been made and this is ready to go in.

priley86 commented:
Filed #530 for any follow-up enhancements. This looks good to go from my standpoint!
